File: tui/src/scala/tui/Spans.scala
package tui

/// A string composed of clusters of graphemes, each with their own style.
case class Spans(spans: Array[Span]) {
  /// Returns the width of the underlying string.
  def width: Int =
    spans.map(_.width).sum

  override def toString: String = spans.mkString("")

  /// Apply a new style to existing text.
  def patchedStyle(style: Style, overwrite: Boolean): Spans =
    Spans(spans.map(_.patchedStyle(style, overwrite)))

  def /(style: Style): Spans =
    patchedStyle(style, overwrite = true)

  def /(optionalStyle: Option[Style]): Spans =
    optionalStyle match {
      case Some(style) => this / style
      case None        => this
    }
}

object Spans {
  def nostyle(s: String): Spans = from(Span.nostyle(s))
  def styled(s: String, style: Style): Spans = from(Span.styled(s, style))
  def from(spans: Span*): Spans = Spans(spans.toArray)
}

File: tui/src/java/tui/Spans.java
package tui;

import java.util.Arrays;

/// A string composed of clusters of graphemes, each with their own style.
public record Spans(Span[] spans) {

  /// Returns the width of the underlying string.
  public int width() {
    int sum = 0;
    for (Span s : spans) {
      sum += s.width();
    }
    return sum;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    for (Span s : spans) {
      sb.append(s);
    }
    return sb.toString();
  }

  @Override
  public boolean equals(Object obj) {
    return obj instanceof Spans s && Arrays.equals(spans, s.spans);
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(spans);
  }

  public static Spans nostyle(String s) {
    return from(Span.nostyle(s));
  }

  public static Spans styled(String s, Style style) {
    return from(Span.styled(s, style));
  }

  public static Spans from(Span... spans) {
    return new Spans(spans);
  }

  /// Returns a new Spans whose every Span has its style patched with the given style.
  public Spans patchStyle(Style other) {
    Span[] out = new Span[spans.length];
    for (int i = 0; i < spans.length; i++) {
      out[i] = spans[i].patchStyle(other);
    }
    return new Spans(out);
  }

  /// Returns a new Spans whose every Span has its style reset to Style.RESET.
  public Spans resetStyle() {
    return patchStyle(Style.RESET);
  }
}
